### DISCUSSION TITLE
save_data incorrect behaviour

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#199] [Design & Development]
Trying to again run the channel-by-channel download from iris with obspy get _waveforms.  For the record, here is the relevant block of code being run:
```
from obspy.clients.fdsn import RoutingClient
from obspy import UTCDateTime
from mspasspy.util.converter import Trace2TimeSeries
client=RoutingClient('iris-federator')
from time import time
# This time range yields a cursor in a query that will drive processing. 
#  that cursor must be immortal
ts=UTCDateTime('2012-01-01T00:00:00.0')
te=UTCDateTime('2012-01-05T00:00:00.0')
#te=UTCDateTime('2012-02-01T00:00:00.0')
query={'time' : {'$gte' : ts.timestamp,
      '$lt' : te.timestamp}}
n=db.arrival.count_documents(query)

# This sets up log files
fwstimes=open('webservice_timing.txt','w')
fdbtimes=open('mongodb_timing.txt','w')
ferrs=open('problem_arrivals_docs.txt','w')
print('Number arrivals to be processed this run=',n)

# This could be made a function with these passed as parameters but for now we freeze these.
# They define times before and after arrival time for data download
dt_start = -300.0
dt_end = +300.0
cursor=db.arrival.find(query,no_cursor_timeout=True)
#cursor=db.arrival.find(query,limit=4)
total_download_time=0.0
total_db_time=0.0
for doc in cursor:
    if 'net' in doc:
        net=doc['net']
    else:
        # This will be passed to obspy get_waveform as a wildcard if net is not defined
        doc="*"
    if 'sta' in doc:
        sta=doc['sta']
    else:
        ferrs.write(str(doc))
        print('Skipping doc with missing sta key (should not happen)')
        continue
    
    # we have to assume time is defined - we'll let this abort if it isn't
    arrival_time=doc['time']
    winstart=arrival_time+dt_start
    winend=arrival_time+dt_end
    print('Working on net=',net,' station=',sta,' arrival time=',UTCDateTime(arrival_time))
    t0=time()
    strm=client.get_waveforms(
        starttime=UTCDateTime(winstart),
        endtime=UTCDateTime(winend),
        network=net,
        station=sta,
        channel='BH?',
        location='*'
    )
    t1=time()
    dt=t1-t0
    fwstimes.write(str(dt)+'\n')
    total_download_time += dt
    if not 'source_id' in doc:
        ferrs.write(str(doc))
    #print('Number of channels for station ',sta,'=',len(strm))
    if len(strm)>3:
        print(net,':',sta,' at ',UTCDateTime(arrival_time),' yielded ',len(strm),' waveforms - watch or dups')
    for tr in strm:
        t0=time()
        d=Trace2TimeSeries(tr)  
        if 'source_id' in doc:
            d['source_id']=doc['source_id']
        aid=doc['_id']
        d['arrival_id']=str(aid)
        print(d)
        db.save_data(d)
        t1=time()
        dt=t1-t0
        total_db_time+=dt
        fdbtimes.write(str(dt)+'\n')
fwstimes.close()
fdbtimes.close()
ferrs.close()
print('Total time to download data=',total_download_time) 
print('Total mspass conversion and dbsave time=',total_db_time)
```
First time I tried to run that all the seed net, sta, chan, and loc keys (required for this to be valid without being rigid) recall the previous version was dropping net, sta, chan and loc completely.  Actually, I think it was worse but that is not relevant.  The problem is I now get this behavior:
```
Number of TimeSeries objects currently in db= 678
first 6 wf_miniseed documents
{'_id': ObjectId('609bd7b269f2946c7774e193'), '_fdsnws_dataselect_url': 'http://service.iris.edu/fdsnws/dataselect/1/query', '_format': 'MSEED', 'arrival_id': '6010126b4b4f9e654b6d9b85', 'calib': 1.0, 'READONLYERROR_chan': 'BHE', 'delta': 0.025, 'endtime': 1325378853.9999979, 'mseed': {'dataquality': 'M', 'number_of_records': 7, 'encoding': 'STEIM2', 'byteorder': '>', 'record_length': 4096, 'filesize': 81920}, 'READONLYERROR_net': 'TA', 'npts': 24000, 'sampling_rate': 40.0, 'source_id': '6012d95c961f396071be8657', 'READONLYERROR_sta': 'I03D', 'starttime': 1325378254.024998, 'time_standard': 'UTC', 'utc_convertible': True, 'elog_id': ObjectId('609bd7b269f2946c7774e192'), 'gridfs_id': ObjectId('609bd7b269f2946c7774e194'), 'storage_mode': 'gridfs'}
{'_id': ObjectId('609bd7b269f2946c7774e197'), '_fdsnws_dataselect_url': 'http://service.iris.edu/fdsnws/dataselect/1/query', '_format': 'MSEED', 'arrival_id': '6010126b4b4f9e654b6d9b85', 'calib': 1.0, 'READONLYERROR_chan': 'BHN', 'delta': 0.025, 'endtime': 1325378853.9999979, 'mseed': {'dataquality': 'M', 'number_of_records': 6, 'encoding': 'STEIM2', 'byteorder': '>', 'record_length': 4096, 'filesize': 81920}, 'READONLYERROR_net': 'TA', 'npts': 24000, 'sampling_rate': 40.0, 'source_id': '6012d95c961f396071be8657', 'READONLYERROR_sta': 'I03D', 'starttime': 1325378254.024998, 'time_standard': 'UTC', 'utc_convertible': True, 'elog_id': ObjectId('609bd7b269f2946c7774e196'), 'gridfs_id': ObjectId('609bd7b269f2946c7774e198'), 'storage_mode': 'gridfs'}
```
I'm not sure where it is coming from but notice net has been rekeyed as READONLYERROR_net with sta and chan being altered the same way.  That is certainly not the right behavior as a default since the default save mode is "promiscuous".   This must be happening the the mapping from the Metadata schema to the wf_TimeSeries schema as the only place I see net, sta, chan, and loc defined as readonly attributes is there.  I am not sure how to fix this.  

I'm going to let my download script run for the full 2012 data set because I know I can fix this problem with the new clean methods (rename option).   Hoping one of you guys could figure out quickly how to fix this issue.  I think you already have the notebook with the above script to work with.

**Comment by wangyinz** [comment]
Well, I think that is actually the correct behavior. This is because you are using the default schema, `mspass.yaml`, which forces a normalized data model where the keys such as net and chan should only be defined in the site collection. Thus, it is readonly for saving to wf. This schema is designed for users doing data processing jobs with a clean database. Here it does not fit for you because you are actually constructing the database itself. So, the fix here is to just switch to our `mspass_lite.yaml` schema, which has the least amount of constraint and should serve the purpose here.

**Reply by pavlis** [reply]
I wondered if that was the solution.   I'm not 100% sure that is the desired behaviour.  To me any insertion of a new document derived from a datum's metadata should not enforce anything about the schema by default.  I'll not argue too strongly about this point BUT we have to make this very very clear in the documentation.

An important documentation page we need that is currently not really complete is a user's manual page on the schema usage and how to extend the schema.   I would suggest strongly you look through [this section](https://wangyinz.github.io/mspass/user_manual/database_concepts.html) and add some material about the standard schemas and a section on extending the schema to extend the package.  

Related issue about the schema.  The user's manual needs a table generated from the yaml file that is an update of the table we had in an earlier time.  I don't remember how that was done before.

**Reply by wangyinz** [reply]
I think the behavior is fine here because read-only is probably the only thing that should always be forced in a schema. If a user writes to a read-only key, it should be treated as an error even in the promiscuous mode. This can avoid resulting in confusing database contents. For example, in `mspass.yaml` we have defined in the metadata schema that the key `sta` should be read from the `site` collection. If we allow user to write a different `sta` into the `wf` collection, we may need to face two conflicting `sta` when read. I am not even sure which `sta` will take precedence in our reader, but either way, it is confusing and error prone. For those that don't want `sta` to be read only, they should probably just use a different schema, which can be the `mspass_lite.yaml` in this case, where no normalization is forced and there is no such confusion.

Yes, I am aware of the schema table issue. This requires re-implement [this script](https://github.com/wangyinz/mspass/blob/master/docs/source/mspass_schema/build_metadata_tbls.py) with the new schema class. I just never got the chance to do so since it is not a simple change.
